src/palace/manager/api/odl/api.py

```diff
@@ -300,46 +221,42 @@ def checkin(self, patron: Patron, pin: str, licensepool: LicensePool) -> None:
 
     def _checkin(self, loan: Loan) -> bool:
         _db = Session.object_session(loan)
-        doc = self.get_license_status_document(loan)
-        status = doc.get("status")
-        if status in [
-            self.REVOKED_STATUS,
-            self.RETURNED_STATUS,
-            self.CANCELLED_STATUS,
-            self.EXPIRED_STATUS,
-        ]:
-            # This loan was already returned early or revoked by the distributor, or it expired.
-            self.update_loan(loan, doc)
-            raise NotCheckedOut()
+        if loan.external_identifier is None:
+            # We can't return a loan that doesn't have an external identifier. This should never happen
+            # but if it does, we log an error and delete the loan, so it doesn't stay on the patrons
+            # bookshelf forever.
+            self.log.error(f"Loan {loan.id} has no external identifier.")
+            return False
 
-        return_url = None
-        links = doc.get("links", [])
-        for link in links:
-            if link.get("rel") == "return":
-                return_url = link.get("href")
-                break
+        doc = self._request_loan_status("GET", loan.external_identifier)
+        if not doc.active:
+            self.log.warning(
+                f"Loan {loan.id} was already returned early, revoked by the distributor, or it expired."
+            )
+            self.update_loan(loan, doc)
+            return False
```

Comment:
This is a functionality change. We used to raise `NotCheckedOut` in this case.

I think the updated behavior is correct. If you are trying to return an already returned item, we should just process the return and delete the local loan, instead of raising an exception and keeping the loan on the users shelf, in a state where it cannot be returned.